src/chainstate/stacks/index/proofs.rs

```diff
@@ -0,0 +1,1129 @@
+    /// them.  That is, make a proof that the current trie's root node hash and ptr are only reachable from the
+    /// corresponding non-backptr root in this trie's ${ptr.back_block()}th ancestor back.
+    /// s must point to the block from which we're going to walk back from.
+    /// 
+    /// The first entry of the shunt proof is the set of Trie root hashes _excluding_ the one from
+    /// backptr, as well as the index into the list of Trie root hashes into which the backptr hash
+    /// should be inserted (this root hash is calculated from the segment proof for that backptr
+    /// node).
+    ///
+    /// The last entry of the shunt proof is the set of root hashes _excluding_ the final root
+    /// hash, which will be the root hash for the segment proof for the non-backptr copy of this
+    /// node.
+    ///
+    /// All intermediate shunt proofs will contain all ancestor hashes for each node in-between the
+    /// backptr and the non-backptr node.  The intermediate root hashes will be calculated by the verifier.
+    fn make_shunt_proof(storage: &mut TrieFileStorage, backptr: &TriePtr) -> Result<Vec<TrieMerkleProofType>, Error> {
+        // the proof is built "backwards" -- starting from the current block all the way back to backptr.
+        // Note that it is okay if backptr is not an actual backptr
+        let mut proof = vec![];
+        
+        let mut back_block = backptr.back_block();
+        let mut block_header = storage.get_cur_block();
+
+        let ancestor_block_hash = storage.block_walk(back_block)?;
+        storage.open_block(&ancestor_block_hash, false)?;
+
+        let ancestor_root_hash = read_root_hash(storage)?;
+
+        let mut found_backptr = false;
+        
+        // find last and intermediate entries in the shunt proof -- exclude the root hashes; just
+        // include the ancestor hashes.
+        while back_block > 0 && !found_backptr {
+            storage.open_block(&block_header, false)?;
+            let cur_root_hash = read_root_hash(storage)?;
+            trace!("Shunt proof: walk back {} from {:?} ({:?})", back_block, &block_header, &cur_root_hash);
+
+            let mut ancestor_hash_buf = Vec::with_capacity(TRIEHASH_ENCODED_SIZE * 256);
+            Trie::get_trie_ancestor_hashes_bytes(storage, &mut ancestor_hash_buf)?;
+
+            let ancestor_hashes = hash_buf_to_trie_hashes(&ancestor_hash_buf);
+            trace!("Ancestors of {:?} ({:?}): {:?}", &block_header, &cur_root_hash, &ancestor_hashes);
+
+            // did we reach the backptr's root hash?
+            for i in 0..ancestor_hashes.len() {
+                if ancestor_hashes[i] == ancestor_root_hash {
+                    found_backptr = true;
+                    break;
+                }
+            }
+
+            // what's the next block we'll shunt to?
+            let mut idx = 0;
+            for i in 0..ancestor_hashes.len() {
+                if (1u32 << i) <= back_block {
```

Comment:
These skip lists look ups (i.e., what's the next index in the skip list) should probably be functions of the MARF struct to ensure that backwards iteration over the list is the same everywhere.

Reply:
I can do that.

Reply:
Actually, looking at this deeper, it's not clear to me what would be gained by moving this to the MARF struct.  The MARF never needs to use the skip-list to look up prior nodes -- it just uses the fork table within the `TrieFileStorage` instance.  The MARF only needs to _write_ the skip-list on calls to `update_root_hash()`.  Only the prover needs to read the skip-list, and trying to move the logic to do so away from the prover I think would make it harder for the reader to understand the verifier.